src/Jackalope/Version/VersionHandler.php

```diff
@@ -235,15 +236,203 @@ private function createFrozenNode(NodeInterface $versionNode, NodeInterface $nod
+        // also see (https://github.com/jackalope/jackalope/issues/247)
+
+        if ($frozenNode->hasProperty('jcr:frozenMixinTypes')) {
+            $node->setProperty('jcr:mixinTypes', $frozenNode->getPropertyValue('jcr:frozenMixinTypes'));
+        } elseif ($node->hasProperty('jcr:mixinTypes')) {
+            $node->getProperty('jcr:mixinTypes')->remove();
+        }
+
+        // handle properties present on the frozen node
+        foreach ($frozenNode->getProperties() as $property) {
+            /** @var PropertyInterface $property */
+            $propertyName = $property->getName();
+            if ($propertyName == 'jcr:frozenPrimaryType'
+                || $propertyName == 'jcr:frozenMixinTypes'
+                || $propertyName == 'jcr:frozenUuid'
+            ) {
+                continue;
+            }
+
+            if ($node->hasProperty($propertyName)) {
+                $nodeProperty = $node->getProperty($propertyName);
+                $onParentVersion = $nodeProperty->getDefinition()->getOnParentVersion();
+                if ($onParentVersion == OnParentVersionAction::COPY || $onParentVersion == OnParentVersionAction::VERSION) {
+                    $nodeProperty->setValue($property->getValue());
+                }
+            } else {
+                // cannot check the onParentVersion attribute of a non existing property
+                // but if the property exists on the frozen node it has to be copy or version
+                $node->setProperty($propertyName, $property->getValue());
+            }
+        }
+
+        // handle properties present on the node but not on the frozen node
+        foreach ($node->getProperties() as $propertyName => $property) {
+            if ($frozenNode->hasProperty($propertyName)) {
+                continue;
+            }
+
+            $onParentVersion = $property->getDefinition()->getOnParentVersion();
+
+            switch ($onParentVersion) {
+                case OnParentVersionAction::COPY:
+                case OnParentVersionAction::VERSION:
+                case OnParentVersionAction::ABORT:
+                    $property->remove();
+                    break;
+                case OnParentVersionAction::INITIALIZE:
+                case OnParentVersionAction::COMPUTE: // implementation-specific, decided to handle as INITIALIZE
+                    // TODO how to get default value of property?
+            }
+        }
+
+        // TODO handle identifier collisions
+
+        // handle child nodes present on the frozen node
+        foreach ($frozenNode->getNodes() as $frozenChildNode) {
+            /** @var NodeInterface $frozenChildNode */
+            if (!$removeExisting) {
+                // TODO check occurence of node in repository
+                throw new NotImplementedException('Check for $removeExisting not implemented yet');
+            }
+
+            // TODO handle different onParentVersion values
```

Comment:
You were right, the children have also the node type `nt:frozenNode`... But that doesn't make my issue easier to solve. I think that means that the logic from `getDefinition` has to be duplicated in some way, because the one existing does not work with all the frozen* properties, or am I missing something?